File: Lab2_13_2/Lab2_13_2/PhoneCall.cs
using System;

namespace Lab2_13_2
{
    public class PhoneCall
    {
        public string PhoneNumber { get; set; }

        public string TimeOfStartOfPhoneCall { get; }

        public string TimeOfEndOfPhoneCall { get; }

        public int PhoneCallLength { get; }

        public decimal PhoneCallCost { get; } // -> new segregated class

        private int TimeOfStartOfPhoneCallInMinutes; // -> new segregated class

        private int TimeOfEndOfPhoneCallInMinutes; // -> new segregated class

        private int MinutesOnLowestTariff; // -> new segregated class

        private int MinutesOnHighestTariff; // -> new segregated class


        public PhoneCall(string number, string start, string end)
        {
            PhoneNumber = number;
            TimeOfStartOfPhoneCall = start;
            TimeOfEndOfPhoneCall = end;
            TimeOfStartOfPhoneCallInMinutes = ConvertInMinutes(TimeOfStartOfPhoneCall);
            TimeOfEndOfPhoneCallInMinutes = ConvertInMinutes(TimeOfEndOfPhoneCall);
            PhoneCallLength = GetLengthOfPhoneCallInMinutes(TimeOfStartOfPhoneCallInMinutes, TimeOfEndOfPhoneCallInMinutes);
            MinutesOnHighestTariff = CountMinutesOnHighestTariff();
            MinutesOnLowestTariff = CountMinutesOnLowestTariff();
            PhoneCallCost = GetCostOfPhoneCall();
        }

        public decimal GetCostOfPhoneCall()
        {

            const decimal COST_OF_PHONE_CALL_BETWEEN_9_20 = 1.50M;
            const decimal COST_OF_PHONE_CALL_REGULAR = 0.90M;

            return MinutesOnHighestTariff * COST_OF_PHONE_CALL_BETWEEN_9_20 + MinutesOnLowestTariff * COST_OF_PHONE_CALL_REGULAR;
        }

        private int CountMinutesOnHighestTariff()
        {

            int NineAMInMinutes = ConvertInMinutes("9:00");
            int EigthPMInMinutes = ConvertInMinutes("20:00");

            if (TimeOfStartOfPhoneCallInMinutes < NineAMInMinutes)
            {
                if (TimeOfEndOfPhoneCallInMinutes < NineAMInMinutes)
                {
                    return 0;
                }
                else if (TimeOfEndOfPhoneCallInMinutes >= NineAMInMinutes && TimeOfEndOfPhoneCallInMinutes <= EigthPMInMinutes)
                {
                    return TimeOfEndOfPhoneCallInMinutes - NineAMInMinutes;
                }
                else
                {
                    return GetLengthOfPhoneCallInMinutes(NineAMInMinutes, EigthPMInMinutes);
                }

            }
            else if (TimeOfStartOfPhoneCallInMinutes >= NineAMInMinutes && TimeOfStartOfPhoneCallInMinutes <= EigthPMInMinutes)
            {
                if (TimeOfEndOfPhoneCallInMinutes >= NineAMInMinutes && TimeOfEndOfPhoneCallInMinutes <= EigthPMInMinutes)
                {
                    return PhoneCallLength;
                }
                else
                {
                    return EigthPMInMinutes - TimeOfStartOfPhoneCallInMinutes;
                }
            }
            else
            {
                return 0;
            }
        }

        private int CountMinutesOnLowestTariff()
        {
            return PhoneCallLength - MinutesOnHighestTariff;
        }

        private int GetLengthOfPhoneCallInMinutes(int intervalStart, int intervalFinish)
        {
            return 1440 - intervalStart - (1440 - intervalFinish);
        }

        private int ConvertInMinutes(string time)
        {
            string[] values = time.Split(":");
            return Convert.ToInt32(values[0]) * 60 + Convert.ToInt32(values[1]);
        }

        private bool IsAmountOfCostNonNegative()
        {
            return PhoneCallCost >= 0;
        }

        public override string ToString()
        {
            if (IsAmountOfCostNonNegative())
            {
                return "Phone number: " + PhoneNumber +
                "; \nStart of Phone Call: " + TimeOfStartOfPhoneCall +
                "; \nEnd of Phone Call: " + TimeOfEndOfPhoneCall +
                "; \nCost of Phone Call: " + PhoneCallCost +
                "; \nLength of Phone Call in Minutes: " + PhoneCallLength +
                ";";
            }
            else
            {
                return "Not appropriate input data: Phone call length cannot be less than 0 minutes";
            }
        }
    }
}

File: Lab2_13_2/Lab2_13_2_Python/PhoneCall.py
class PhoneCall():

    def __init__(self, phone_number: str, start: str, end: str) -> None:
        self.__phone_number = phone_number
        self.__start_of_phone_call = start
        self.__end_of_phone_call = end

    def get_cost_of_phone_call(self) -> float:
        COST_OF_PHONE_CALL_BETWEEN_9_20 = 1.50
        COST_OF_PHONE_CALL_REGULAR = 0.90
        return self.__count_minutes_on_highest_tariff() * COST_OF_PHONE_CALL_BETWEEN_9_20 + self.__count_minutes_on_lowest_tariff() * COST_OF_PHONE_CALL_REGULAR

    def __count_minutes_on_highest_tariff(self) -> int:
        nine_AM_in_minutes = self.__convert_in_minutes("9:00")
        eigth_PM_in_minutes = self.__convert_in_minutes("20:00")
        start_of_phone_call_in_minutes = self.__convert_in_minutes(
            self.__start_of_phone_call)
        end_of_phone_call_in_minutes = self.__convert_in_minutes(
            self.__end_of_phone_call)

        if(start_of_phone_call_in_minutes < nine_AM_in_minutes):
            if(end_of_phone_call_in_minutes < nine_AM_in_minutes):
                return 0
            elif(end_of_phone_call_in_minutes >= nine_AM_in_minutes and end_of_phone_call_in_minutes <= eigth_PM_in_minutes):
                return end_of_phone_call_in_minutes - nine_AM_in_minutes
            else:
                return self.__get_length_of_phone_call_in_minutes(nine_AM_in_minutes, eigth_PM_in_minutes)
        elif(start_of_phone_call_in_minutes >= nine_AM_in_minutes and start_of_phone_call_in_minutes <= eigth_PM_in_minutes):
            if(end_of_phone_call_in_minutes >= nine_AM_in_minutes and end_of_phone_call_in_minutes <= eigth_PM_in_minutes):
                return self.__get_length_of_phone_call_in_minutes(start_of_phone_call_in_minutes, end_of_phone_call_in_minutes)
            else:
                return eigth_PM_in_minutes - start_of_phone_call_in_minutes
        else:
            return 0

    def __count_minutes_on_lowest_tariff(self) -> int:
        start_of_phone_call_in_minutes = self.__convert_in_minutes(
            self.__start_of_phone_call)
        end_of_phone_call_in_minutes = self.__convert_in_minutes(
            self.__end_of_phone_call)
        return self.__get_length_of_phone_call_in_minutes(start_of_phone_call_in_minutes, end_of_phone_call_in_minutes) - self.__count_minutes_on_highest_tariff()

    def __get_length_of_phone_call_in_minutes(self, interval_start: int, interval_end: int) -> int:
        return 1440 - interval_start - (1440 - interval_end)

    def __convert_in_minutes(self, time: str) -> int:
        values = time.split(":")
        return int(values[0]) * 60 + int(values[1])

    def __is_amount_of_cost_non_negative(self) -> bool:
        return self.get_cost_of_phone_call()

    def get_current_phone_call_length(self) -> int:
        return self.__get_length_of_phone_call_in_minutes(
            self.__convert_in_minutes(
                self.__start_of_phone_call),
            self.__convert_in_minutes(self.__end_of_phone_call))

    def __repr__(self) -> str:
        if(self.__is_amount_of_cost_non_negative()):
            return "\nPhone number: " + self.__phone_number + "; \nStart of Phone Call: " + self.__start_of_phone_call + "; \nEnd of Phone Call: " + self.__end_of_phone_call + "; \nCost of Phone Call: " + str(self.get_cost_of_phone_call()) + "; \nLength of Phone Call in Minutes: " + str(self.get_current_phone_call_length()) + ";\n"
        return "\nNot appropriate input data: Phone call length cannot be less than 0 minutes"
